### PR TITLE
Zoltan2 directory

### PR DESCRIPTION
**Pull request by MicheldeMessieres**
Note this PR is in progress and not ready yet for merging - setting this up to create commenting opportunities.

In the current state the unit test for the Kokkos directory demonstrates working Add and Replace modes for variable sized arrays. Aggregate mode will be added soon.

To get this working I implemented some code for the usage case which was previously marked as unsupported: _Non-blocked, variable-sized recvs not supported_. Some critical areas for discussion are marked with the text NEW METHOD. Even though this works I suspect this may not cover all usage cases or be the best approach and would like to discuss that further. In particular I have the do_wait method filling the receive buffer out of order (because the directory handles resorting anyways) but this is not consistent with the original implementation.

### CODE REVIEW

**Comment by mhoemmen** [comment]
@MicheldeMessieres May I ask how this differs from Tpetra's existing Directory?  Is this another implementation of the same idea?  Were there performance issues in Tpetra that required a clean-slate implementation?

**Comment by kddevin** [comment]
@mhoemmen In April, you and I discussed the use-cases needed for Zoltan2.  This work addresses those use cases.

**Comment by mhoemmen** [comment]
@kddevin Thanks for clarifying!  It could still be worthwhile to compare against Tpetra performance.  Tpetra's Directory is self-contained enough that it may not be hard to replace if you manage to improve on it.

**Comment by MicheldeMessieres** [comment]
@kddevin @mndevec Updated this PR with various items we had discussed for the directory. Will send further notes by e-mail.

File: packages/zoltan/src/Utilities/DDirectory/zoltan_dd_const.h [hunk 1]
```diff
@@ -47,6 +47,13 @@
 #ifndef ZOLTAN_DD_H
 #define ZOLTAN_DD_H
 
+// A temporary addition for profiling
+// This will be removed in the final version
+#include <time.h>
+#define GET_TIME ((double)(clock())/(double)(CLOCKS_PER_SEC))
```
Comment (kddevin) [inline packages/zoltan/src/Utilities/DDirectory/zoltan_dd_const.h]: Can we remove the timers from the Zoltan directory files?

File: packages/zoltan2/src/algorithms/partition/Zoltan2_TaskMapping.hpp [hunk 2]
```diff
@@ -28,6 +28,10 @@
 //#define gnuPlot
 #include "Zoltan2_XpetraMultiVectorAdapter.hpp"
 
+// TODO: Delete this define once other temporary development modes are removed
+#define CONVERT_DIRECTORY_KOKKOS
```
Comment (kddevin) [inline packages/zoltan2/src/algorithms/partition/Zoltan2_TaskMapping.hpp]: Comment says this define should be deleted.

File: packages/zoltan2/src/directory/Zoltan2_Directory.hpp [hunk 3]
```diff
@@ -0,0 +1,791 @@
...
+    int owner;       /* processor hosting global ID object    */
+    int partition;   /* Optional data                         */
+    int errcheck;    /* Error checking(inconsistent updates)  */
+    lid_t lid;       /* lid value */
+    user_t userData; /* user data */
+    int free;        /* flag whether node is free or used     */
+};
+#endif
+
+#ifdef CONVERT_DIRECTORY_RELIC
```
Comment (kddevin) [inline packages/zoltan2/src/directory/Zoltan2_Directory.hpp]: Should CONVERT_DIRECTORY_RELIC code now be removed?

File: packages/zoltan2/test/directory/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+TRIBITS_ADD_EXECUTABLE_AND_TEST(
+  directoryTest_KokkosSimple
+  SOURCES directoryTest_KokkosSimple.cpp
+  NUM_MPI_PROCS 4
+  COMM serial mpi
+)
+
+# Temporary tests for development and comparison
+# with above Kokkos tests. To be deleted.
```
Comment (kddevin) [inline packages/zoltan2/test/directory/CMakeLists.txt]: Which tests should we keep?

**Comment by MicheldeMessieres** [comment]
@kddevin I'll proceed to pull out all the profiling code and the 3 unused modes. I will also remove the 'Kokkos' mode #define and leave that code as the default. I'll remove any tests related to that. If we need to do any further profiling we could just compare to develop which is probably best anyways. That should be a quick set of edits.

**Comment by MicheldeMessieres** [comment]
@kddevin I left the new test for findUniqueGids using the new directory. This is in directoryTest_findUniqueGids.cpp. We may want to remove the original test and replace it with this one but will wait on that until we discuss.

**Comment by mhoemmen** [comment]
@MicheldeMessieres @kddevin The usual way to fix ANSI strict aliasing warnings is to use `memcpy` instead of direct assignment.  For example:
```c++
memcpy (&lid[ptr->index], ptr->adjData, sizeof (lid_t));
```

**Comment by MicheldeMessieres** [comment]
@mhoemmen  Thanks for the tip on the strict alias warnings.
@kddevin These warnings are all fixed now.

**Pull request merged**